base/intfuncs.jl

```diff
@@ -314,13 +315,13 @@ to_power_type(x) = convert(Base._return_type(*, Tuple{typeof(x), typeof(x)}), x)
     "or write float(x)^", p, " or Rational.(x)^", p, ".")))
 # The * keyword supports `*=checked_mul` for `checked_pow`
 @assume_effects :terminates_locally function power_by_squaring(x_, p::Integer; mul=*)
-    x = to_power_type(x_)
+    (x, x2) = promote(x_, mul(x_, x_))
```

Comment:
This is close to how it used to be, but de7f5ae changed it to avoid the unnecessary allocation for matrices.

Reply:
I don't think that having the behavior vary depending on input type is such a good idea. Especially because the code is quite inefficient allocation-wise for matrix input anyway.

Reply:
Looks like it's necessary to avoid the promotion for matrices in any case, looking at the failing tests.